### DISCUSSION TITLE
Resources representation

### DISCUSSION THREAD

**Discussion by mikand** [aiplan4eu/unified-planning#44] [Ideas]
In WP3 meetings we discussed several options for implementing a model of resources int he UPF that abstracts away from the "manual" encoding into numeric planning. This issue is intended to summarize the discussion so far and to keep track of proposals, ideas and snippets.

### Objective
Instead of manually modeling numerical quantities and their dynamics in the action conditions and effects, provide high-level structures to model the resource and its dynamics. This is to simplify and avoid reinventing the wheel for the user, but also to bridge the gap with scheduling and timeline-based planning and allow for future planners with dedicated resources support.

### Proposal spirit
We want to allow for resources to be instantiated from a set of templates (e.g. "RealInstantaneousConsumableResource") taht are defined in the UPF and have a semantic explanation attached to them. Resources are added as first-class citizens in the problem, but the UPF will provide support for compiling them away in numeric/hybrid planning. In this way a solver can either handle resources natively or compile them away and use more general techniques.

### Proposal Example
This is an example implementation of a `RealInstantaneousConsumableResource`, that is a real-valued resourced that can be consumed or produced instantaneously.

This code is on the UPF side:
```python
class InstantaneousRealConsumableResource(Resource):
    '''
    This Resource models a real-valued quantity that is consumed (or produced) instantaneously
   
    E.g. A  battery can be modeled as follows.
    Suppose the move(from, to) action consumes distance(from, to) * consumption_rate battery units when applied 
      InstantaneousRealConsumableResource('battery', 
                                          {move: Times(duration(from, to), consumption_rate}, 
                                          100)
    '''
    def __init__(self, name, consuming_actions=None, initial_value=None):
        self.consuming_actions = consuming_actions
        self.name = name
        self.initial_value = initial_value

    def add_consuming_action(self, action, consumption_expression):
        '''
        Add a new action taht consumes or produces this resource.
        `consumption_expression` is an expression modeling how much resource is produced/consumed
        '''
        self.consuming_actions[action] = consumption_expression


def InstantaneousRealConsumableResourceCompiler(problem, resource):
    '''
    This method modifies `problem` to model the dynamics of the `resource`.
    '''
    domain = upf.Fluent(resource.name, RealType())
    problem.add_fluent(domain)

    for a in problem.actions():
        consume = resource.consuming_actions.get(a, None)
        if consume is not None:
            a.add_precondition(GE(domain, consume))
            a.add_decrease_effect(domain, consume)
    problem.set_initial_value(domain, resource.initial_value)
``` 

On the user side, a modeling like this can be used as follows:
```python
Location = UserType('Location')
robot_at = upf.Fluent('robot_at', BoolType(), [Location])
distance = upf.Fluent('distance', RealType(), [Location, Location])
move = upf.Action('move', l_from=Location, l_to=Location)
l_from = move.parameter('l_from')
l_to = move.parameter('l_to')
# move.add_precondition(GE(battery, 10)) # This is created automatically
move.add_precondition(Not(Equals(l_from, l_to)))
move.add_precondition(robot_at(l_from))
move.add_precondition(Not(robot_at(l_to)))
move.add_effect(robot_at(l_from), False)
move.add_effect(robot_at(l_to), True)
# move.add_effect(battery.decrease(10)) -> This is done automagically
l1 = upf.Object('l1', Location)
l2 = upf.Object('l2', Location)
problem = upf.Problem('robot')
problem.add_fluent(robot_at)
battery = InstantaneousRealConsumableResource('battery', {move: distance(l_from, l_to)}, initial_value=100)
problem.add_resource(battery)
problem.add_action(move)
problem.add_object(l1)
problem.add_object(l2)
problem.set_initial_value(robot_at(l1), True)
problem.set_initial_value(robot_at(l2), False)
# problem.set_initial_value(battery, 100) # this calls the `get_initial_state_assignment` method
problem.add_goal(robot_at(l2))
```

To maintain the definition of the resource closer to the Acton consuming it we could do instead:
```python
Location = UserType('Location')
robot_at = upf.Fluent('robot_at', BoolType(), [Location])
distance = upf.Fluent('distance', RealType(), [Location, Location])
battery = InstantaneousRealConsumableResource('battery', initial_value=100)
move = upf.Action('move', l_from=Location, l_to=Location)
l_from = move.parameter('l_from')
l_to = move.parameter('l_to')
battery.add_consuming_action(move, distance(l_from, l_to))
move.add_precondition(Not(Equals(l_from, l_to)))
move.add_precondition(robot_at(l_from))
move.add_precondition(Not(robot_at(l_to)))
move.add_effect(robot_at(l_from), False)
move.add_effect(robot_at(l_to), True)
l1 = upf.Object('l1', Location)
l2 = upf.Object('l2', Location)
problem = upf.Problem('robot')
problem.add_fluent(robot_at)
problem.add_resource(battery)
problem.add_action(move)
problem.add_object(l1)
problem.add_object(l2)
problem.set_initial_value(robot_at(l1), True)
problem.set_initial_value(robot_at(l2), False)
problem.add_goal(robot_at(l2))
```

### Issues / Comments

- It could be confusing to not associate the resource with the actions that consume or produce it. In the first example, having the resource "far away" from the move action can be confusing
- Resources should be parametrized, so that we can create for example one battery for each robot in the problem
- The initial state might be moved to the problem to reuse the same resource over multiple `Problem`s
- Recording actions that increase/decrease or set the resource could be more flexible, but it risks to be harder to deal with in scheduling/timeline
- We shall complement modeled resources with support for **simulated** resources, possibly with the capability of providing an approximation of their behavior

**Comment by arbimo** [comment]
Thanks a lot for the write up @mikand. 
Below is an overview of the proposal I mentioned during the meeting, which aims at making usage closer to normal fluents. A benefit is also that it facilitates the parameterization of resources.

```python
# declaration mimics fluents
upf.ConsumableResource('battery', parameters=[Robot], initial_value=100)
...
# usage mimics effects
move_action.add_consumption(battery(rob), distance(l_from, l_to))
```
Of course, it would be great to be able to iterate over all usages of a resource, but I believe we can provide a method on the `Problem` to do this (and normal fluents would benefit from a similar feature as well).

**Comment by arbimo** [comment]
A quick note on what makes resources special and how they can be handled:

1. resources typically have a minimal and maximal capacity
2. they can be used in discrete or continuous amounts (e.g. available machines vs available fuel)
3. they have an initial value
4. they have different allowed usage patterns (borrow, consume, produce, ...)

I believe that (1.) and (2.) should be part of the definition of the resource. (3.) should be part of the initial state definition.

On the other hand, I am not sure whether we should enforce the allowed usage patterns at definition time. My intuition is that it would lead to an explosion of the number of `Resource` subclasses (e.g. `InstantaneousRealConsumableResource`).
Another practical choice would be to let the type of the resources be implicitly defined by the operations on it. For instance, a resource that is only consumed in all action instances would be assumed to be a strictly consumable action. The downside would be that it allows the domain modeler to misuse a resource.

**Comment by hstairs** [comment]
Thanks @mikand and @arbimo for kicking off the discussion.
As said in the meeting, I am also more inclined to adopt an action-centered perspective and therefore prefer @arbimo's proposal. I would also add to the discussion a couple of other things:

1. An action centered perspective is closer to what a declarative PDDL/ANML representation could look like in the future. I can totally see a new field in the action structure that models what and how much an action consumes a given resource (or more). Maybe we can exploit this for reasoning in a more directed way.
2. We can see resources just as other types of variables. Similarly to what happens with numeric fluents that can be affected using increase, decrease and assignment operations, we would need to define how a resource can be modified. Consumption/production. The bit that we take implicit here is the fact that you can’t consume more than what you have, and you can’t produce more than what you can store. This is basically given by the domain of a resource variable. I am not sure about borrowing. That would probably be more suited when concurrency comes into play as I guess it is there where resources are used as some mechanism for regulating access to a critical section.
3. If we commit to this, we can then define a taxonomy for resources with implicit constraints on their use (eg. can’t be replenished)

**Comment by arbimo** [comment]
Having thought a bit more about this, I believe that what we currently implement for numeric planning is at the right level of detail for resource modeling. For the record, it is essentially something like so (in ANML-like syntax):
```
// bounds must be constants values (ideally we could allow static expressions)
// they are just syntactic sugar for requirements over the entire plan
int[0,100] battery(Robot r); 

[start] battery(R1) := 40; // initial value
[start] battery(R2) := 80;

// further restrictions on allowed values, necessary to distinguish on the requirement of the battery resource for different robots
[start,end] battery(R1) >= 20;
```
The `[0,100]` bounds on the battery fluent capture the right information that it is forbidden to go below/above the given thresholds. (We should make it clear the semantics).

In my perspective the missing parts are:
 - a shorthand to express borrowing patterns that acquire the resource at start and release at end 
 - a way to stick to discrete/integer time semantics (not limited to resource modeling)


The latter appears to be essential to leverage existing scheduling solvers. It also makes it impossible to exactly model the most basic scheduling problems. For instance, a jobshop problem model in the UP would require an epsilon gap between any pair of successive tasks. This is because in planning, resource can only be acquired over the [start,end] interval. On the other hand in scheduling, intervals are typically considered right opened `[start,end)` which allow gap-less sequences of activities.
